course5/homework10/string_with_iterators/string.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <memory>
+#include <ostream>
+
+#include <iostream>
+
+namespace String {
+
+class String
+{
+public:
+	String() = default;
+	~String() = default;
+	String(const String&);
+	String(String&&);
+	String& operator=(const String&);
+	String& operator=(String&&);
+
+	String(const char*);
+	String& operator=(const char*);
+
+	String operator+(const String&);
+	String& operator+=(const String&);
+
+	char& operator[](const size_t idx);
+	const char& operator[](const size_t idx) const;
+
+	bool operator==(const String&) const;
+	bool operator!=(const String& rhs) const { return !(*this == rhs); }
+
+	const char* c_str() const { return mSize > 0 ? mContents.get() : ""; }
+	size_t length() const { return mSize > 0 ? mSize - 1 : 0; }
+
+	template <typename T>
+	class StringIterator:
```

Comment:
Oh no! String stores data in a continous memory block. This means that you can simply use ```char*``` and ```const char*``` as ```iterator``` and ```const_iterator```